proposals/p2274.md

```diff
@@ -0,0 +1,384 @@
+
+Cloned from [#1356](https://github.com/carbon-language/carbon-lang/pull/1356),
+which was cloned from
+[#1059](https://github.com/carbon-language/carbon-lang/pull/1059).
+
+## Problem
+
+Carbon needs a convenient way of indexing into objects like arrays and slices.
+
+## Background
+
+For purposes of this proposal, a "slice" is an object that refers to some region
+of an array, and provides access to it, such as a `std::string_view` or
+`std::span` in C++. Slices have subtly different indexing API semantics than
+arrays: indexing into an array can produce an l-value only if the array itself
+is an l-value, but indexing into a slice can produce an l-value regardless of
+the slice's value category.
+
+Interfaces are Carbon's
+[only static open extension mechanism](/docs/project/principles/static_open_extension.md),
+so user-defined indexing must be defined in terms of interfaces. In order to
+support
+[definition checking of generic functions](/docs/design/generics/goals.md#generics-will-be-checked-when-defined),
+it must be possible to fully typecheck indexing operations using only the
+information exposed by those interfaces. And in order to provide
+[coherence](/docs/design/generics/goals.md#coherence), a given indexing
+expression must execute the same code regardless of how much type information we
+have about it, so long as we have enough type information to establish that the
+expression is valid.
+
+The primary problem this proposal needs to solve is how to support both
+array-like and slice-like indexing semantics while satisfying those
+requirements.
+
+## Proposal
+
+Carbon will support indexing using the conventional `a[i]` subscript syntax.
+When `a` is an l-value, the result of subscripting will always be an l-value,
+but when `a` is an r-value, the result can be an l-value or an r-value,
+depending on which interface the type implements:
+
+-   If subscripting an r-value produces an r-value result, as with an array, the
+    type should implement `IndexWith`.
+-   If subscripting an r-value produces an l-value result, as with a slice, the
+    type should implement `IndirectIndexWith`.
+
+`IndirectIndexWith` is a subtype of `IndexWith`, and subscript expressions are
+rewritten to method calls on `IndirectIndexWith` if the type is known to
+implement that interface, or to method calls on `IndexWith` otherwise.
+
+A type can implement at most one of these two interfaces. `IndirectIndexWith`
+provides a final blanket `impl` of `IndexWith`, which ensures coherence.
+
+One implication of this design is that the requirement of coherence must apply
+to value categories as well as types (and for those purposes, "lvalue" is a
+subcategory of "rvalue"). This implies that for any valid expression containing
+an lvalue sub-expression, replacing that sub-expression with an rvalue
+expression cannot change the validity or semantics of the outer expression, so
+long as the rvalue expression has the same type and evaluates to the same value.
```

Comment:
Is this the right way around? For `x = 1;`, replacing the lvalue expression `x` with an rvalue expression should presumably change validity. Should this be:

```suggestion
subcategory of "rvalue"). This implies that for any valid expression containing
an rvalue sub-expression, replacing that sub-expression with an lvalue
expression cannot change the validity or semantics of the outer expression, so
long as the lvalue expression has the same type and evaluates to the same value.
```

Reply:
Whoops, yes.